### PR TITLE
fix(gltf, tile-converter): attributeStorageInfo, use class name

### PR DESCRIPTION
**Pull request by mspivak-actionengine**
Populate attributeStorageInfo with data according to metadata class name specified

### CODE REVIEW

File: modules/gltf/src/lib/extensions/EXT_structural_metadata.ts [hunk 1]
```diff
@@ -48,42 +48,53 @@ export function getPropertyTableFromExtStructuralMetadata(
...
-
-    return propertyTableWithData;
   }
 
   // eslint-disable-next-line no-console
-  console.warn(
-    'Cannot get property table from EXT_structural_metadata extension. There is neither propertyTables, nor propertyTextures in the extension.'
-  );
+  // console.warn(
+  //   'Cannot get property table from EXT_structural_metadata extension. There is neither propertyTables, nor propertyTextures in the extension.'
```
Comment (ibgreen) [inline modules/gltf/src/lib/extensions/EXT_structural_metadata.ts:76]: Cut? Or add a comment explaining why we keep this commented code.

File: modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts [hunk 2]
```diff
@@ -46,55 +46,53 @@ export function getPropertyTableFromExtFeatureMetadata(
...
-
-      return propertyTable;
-    }
   }
 
   // eslint-disable-next-line no-console
-  console.warn(
-    'Cannot get property table from EXT_feature_metadata extension. There is neither featureTables, nor featureTextures in the extension.'
-  );
+  // console.warn(
```
Comment (ibgreen) [inline modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts:75]: Ditto

**Review by ibgreen — APPROVED** [review]

File: modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts (hunk 2 above)
Comment (mspivak-actionengine) [reply modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts]: Removed.

File: modules/gltf/src/lib/extensions/EXT_structural_metadata.ts [hunk 3]
```diff
@@ -48,42 +48,49 @@ export function getPropertyTableFromExtStructuralMetadata(
   extension: GLTF_EXT_structural_metadata_GLTF,
   metadataClass?: string
 ): FeatureTableJson | null {
+  /**
+   * Note, 3dTiles is able to have multiple featureId attributes and multiple feature tables.
+   * In I3S we should decide which featureIds attribute will be passed to geometry data.
```
Comment (belom88) [inline modules/gltf/src/lib/extensions/EXT_structural_metadata.ts:53]: What the explanation about I3S here? glTF is a lower level module. It doesn't know about I3S and 3DTiles. I would move this function to `tile-converter`.

File: modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts [hunk 4]
```diff
@@ -46,55 +46,49 @@ export function getPropertyTableFromExtFeatureMetadata(
   extension: GLTF_EXT_feature_metadata_GLTF,
   metadataClass?: string
 ): FeatureTableJson | null {
+  /**
+   * Note, 3dTiles is able to have multiple featureId attributes and multiple feature tables.
+   * In I3S we should decide which featureIds attribute will be passed to geometry data.
```
Comment (belom88) [inline modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts:51]: Ditto: remove any mention about I3S and 3DTiles. If the function is used only in `tile-converter`, I would move it there

File: modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts [hunk 5]
```diff
@@ -46,55 +46,49 @@ export function getPropertyTableFromExtFeatureMetadata(
...
-  );
   return null;
 }
 
+/**
+ * Gets data from Feature Table or Feature Texture
+ * @param {GLTF_EXT_feature_metadata_FeatureTable | GLTF_EXT_feature_metadata_FeatureTexture} featureObject
+ * @returns Table containing feature data
+ */
+function getPropertyData<
```
Comment (belom88) [inline modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts:82]: Looks like copy-paste. I believe you can create one generic funciton for `EXT_structural_metadata` and `EXT_feature_metadata`

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 6]
```diff
@@ -1155,39 +1155,41 @@ export default class I3SConverter {
...
    */
   private _convertPropertyTableToNodeAttributes(propertyTable: FeatureTableJson): void {
+    /*
+      We will append attributes if only the attributeStorageInfo doesn't contain anything.
+      If it already has attributes, return from the function.
+      According to ver 1.9 (see https://github.com/Esri/i3s-spec/blob/master/docs/1.9/attributeStorageInfo.cmn.md):
+      "The attributeStorageInfo object describes the structure of the binary attribute data resource of a layer,
+      which is the same for every node in the layer."
+    */
+    if (this.layers0!.attributeStorageInfo?.length) {
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/i3s-converter.ts:1165]: It is the previous version of logic here. The `attributeStorageInfo` is created from first file which has `propertyTable`. That is not correct if we have extensions. I thought about different logic:
* If there is an `EXT_feature_metadata` or `EXT_structural_meatadata` extension, `attributeStorageInfo` is built from the extension JSON metadata. We have there all properties names and datatypes. `attributeStorageInfo` is not updated from `propertyTable` during conversion;
* If there is no any metadata extension, use existing logic, create `attributeStorageInfo` from `propertyTable`, accumulate it from all files, not only from the first one.

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/gltf/src/lib/extensions/EXT_structural_metadata.ts (hunk 3 above)
Comment (mspivak-actionengine) [reply modules/gltf/src/lib/extensions/EXT_structural_metadata.ts]: Done

File: modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts (hunk 4 above)
Comment (mspivak-actionengine) [reply modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts]: Done

File: modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts (hunk 5 above)
Comment (mspivak-actionengine) [reply modules/gltf/src/lib/extensions/deprecated/EXT_feature_metadata.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts (hunk 6 above)
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts [hunk 7]
```diff
@@ -1648,6 +1647,104 @@ export function getPropertyTable(
   }
 }
 
+/**
+ * Handles EXT_structural_metadata to get property table.
+ * @param extension - Global level of EXT_STRUCTURAL_METADATA extension.
+ * @param metadataClass - User selected feature metadata class name.
+ * @returns {FeatureTableJson | null} Property table or null if the extension can't be handled properly.
+ */
+export function getPropertyTableFromExtStructuralMetadata(
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts:1656]: Add a test fot this function

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts [hunk 8]
```diff
@@ -1648,6 +1647,104 @@ export function getPropertyTable(
...
+  return null;
+}
+
+/**
+ * Handles EXT_feature_metadata to get property table.
+ * @param extension - Global level of EXT_FEATURE_METADATA extension.
+ * @param metadataClass - User selected feature metadata class name.
+ * @returns {FeatureTableJson | null} Property table or null if the extension can't be handled properly.
+ */
+export function getPropertyTableFromExtFeatureMetadata(
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts:1692]: Add a test fot this function

File: modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts [hunk 9]
```diff
@@ -102,6 +103,27 @@ export function getAttributeType(key: string, attribute: unknown): string {
   return STRING_TYPE;
 }
 
+/**
+ * Get the attribute type for attributeStorageInfo according to the schema
+ * @see https://github.com/Esri/i3s-spec/blob/master/docs/1.7/attributeStorageInfo.cmn.md
+ * @param {SchemaClassProperty} schema - attribute's schema
+ * @returns attribute's type
+ */
+export function getAttributeTypeBasedOnSchema(schema: SchemaClassProperty): string {
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts:112]: Add a test for the new function

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts (hunk 8 above)
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts:1692]: It's not an exported function anymore. The test is added for getPropertyTable function in geometry-converter.

File: modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts [hunk 10]
```diff
@@ -1648,6 +1647,104 @@ export function getPropertyTable(
...
+  return null;
+}
+
+/**
+ * Gets data from Property Table or Property Texture
+ * @param {
+ *  | GLTF_EXT_structural_metadata_PropertyTable
+ *  | GLTF_EXT_structural_metadata_PropertyTexture
+ *  | GLTF_EXT_feature_metadata_FeatureTable
+ *  | GLTF_EXT_feature_metadata_FeatureTexture
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/geometry-converter.ts:1730]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 11]
```diff
@@ -1151,7 +1163,65 @@ export default class I3SConverter {
...
+    const schemaClasses = this.preprocessData.schemaClasses;
+    let properties: {[key: string]: SchemaClassProperty} | null = null;
+    for (const schemaClass of schemaClasses) {
+      if (schemaClass.classId === this.options.metadataClass) {
+        properties = schemaClass.properties;
+        break;
+      }
+    }
+    if (!properties) {
+      throw new Error(
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 12]
```diff
@@ -1151,7 +1163,65 @@ export default class I3SConverter {
   }
 
   /**
-   * Do conversion of 3DTiles property table to I3s node attributes.
+   * Creates attribute storage info based on the information
+   * collected while analizing extension schemas in 3DTiles.
+   * @return true if attribute storage info is successfully created, false - otherwise.
+   */
+  private _convertSchemaClassesToNodeAttributes(): boolean {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 13]
```diff
@@ -649,14 +653,22 @@ export default class I3SConverter {
...
-        "The attributeStorageInfo object describes the structure of the binary attribute data resource of a layer, which is the same for every node in the layer."
-      */
+    /*
+      First, try to create Attribute storage info based on the schema.
+      The information about a type and a name of attributes has been collected
+      from extension schemas (if any) provided in all tiles so it corresponds to the complete tileset.
+      It's done during the 'analize' stage of the convertion procedure.
+      This information is stored in 'this.preprocessData.schemaClasses'.
+    */
+    const convertedUsingSchema = this._convertSchemaClassesToNodeAttributes();
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 14]
```diff
@@ -1151,7 +1163,65 @@ export default class I3SConverter {
...
+        key,
+        attributeType
+      );
+      const fieldAttributeType = getFieldAttributeType(attributeType);
+      const fieldAttribute = createFieldAttribute(key, fieldAttributeType);
+      const popupInfo = createPopupInfo(Object.keys(propertiesWithObjectId));
+
+      this.layers0!.attributeStorageInfo!.push(storageAttribute);
+      this.layers0!.fields!.push(fieldAttribute);
+      this.layers0!.popupInfo = popupInfo;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts (hunk 9 above)
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts]: Added tests for getAttributePropertiesFromSchema and createStorageAttributes.

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 15]
```diff
@@ -1151,43 +1140,75 @@ export default class I3SConverter {
...
+    let attributePropertySet: AttributePropertySet | null = null;
+    if (this.options.metadataClass) {
+      if (!this.layers0!.attributeStorageInfo!.length && tileContent?.gltf) {
+        attributePropertySet = getAttributePropertiesFromSchema(
+          tileContent.gltf,
+          this.options.metadataClass
+        );
+      }
+    } else {
+      if (propertyTable) {
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/i3s-converter.ts:1166]:
```suggestion
    } else if (propertyTable) {
```

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts [hunk 16]
```diff
@@ -250,3 +257,177 @@ function setupDoubleAttribute(storageAttribute: AttributeStorageInfo): void {
...
+    const storageAttribute: AttributeStorageInfo = createdStorageAttribute(
+      attributeIndex,
+      key,
+      attributeType
+    );
+    const fieldAttributeType = getFieldAttributeType(attributeType);
+    const fieldAttribute = createFieldAttribute(key, fieldAttributeType);
+
+    attributeStorageInfo!.push(storageAttribute);
+    fields!.push(fieldAttribute);
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 17]
```diff
@@ -1151,43 +1140,75 @@ export default class I3SConverter {
...
+        attributePropertySet = this.getAttributePropertyFromPropertyTable(propertyTable);
+      }
+    }
 
-    for (const key in propertyTableWithObjectId) {
+    if (attributePropertySet) {
+      this.layers0!.popupInfo = createStorageAttributes(
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/types.ts [hunk 18]
```diff
@@ -183,12 +183,22 @@ export enum GLTFPrimitiveModeString {
...
 
+export type AttributeProperty = {
+  attributeType: string;
+  attributeDescription?: string;
+  attributeName?: string;
+};
+
+export type AttributePropertySet = {
+  [key: string]: AttributeProperty;
+};
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/types.ts]: Done

File: modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts [hunk 19]
```diff
@@ -250,3 +257,177 @@ function setupDoubleAttribute(storageAttribute: AttributeStorageInfo): void {
...
+ * @param attributeStorageInfo - array where Attribute Storage Info objects are being stored
+ * @param fields - array where attribute fields are being stored
+ * @returns PopupInfo object
+ */
+export function createStorageAttributes(
+  attributePropertySet: AttributePropertySet,
+  attributeStorageInfo: AttributeStorageInfo[] | undefined,
+  fields: Field[] | undefined
+): PopupInfo {
+  const attributesWithObjectId: AttributePropertySet = {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts]: Done

File: modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts [hunk 20]
```diff
@@ -250,3 +257,177 @@ function setupDoubleAttribute(storageAttribute: AttributeStorageInfo): void {
...
+    OBJECTID: {attributeType: OBJECT_ID_TYPE},
+    ...attributePropertySet
+  };
+
+  let attributeIndex = 0;
+  for (const key in attributesWithObjectId) {
+    const attributeProperty = attributesWithObjectId[key];
+    const attributeType = attributeProperty.attributeType;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/feature-attributes.ts]: Done

**Review by belom88 — APPROVED** [review]

**Pull request merged**